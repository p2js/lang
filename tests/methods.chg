type V<T> = { value: T };

impl<T> V<T> {
  fn foo(self*): T* { return &(*self).value; }
}
fn free<T>(s: V<T>*) {}

fn main(): i32 {
  let v: V<i32> = .{ value: 42 };
  free(&v);
  V::<i32>::foo(&v);
  return *(&v).foo();
}
