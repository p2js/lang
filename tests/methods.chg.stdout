fallback @ 14:9 14:11
expr @ 4:33 4:37 => V<T>*
expr @ 4:32 4:37 => V<T>
expr @ 4:32 4:44 => T
expr @ 4:30 4:44 => T*
expr @ 4:23 4:44 => never
expr @ 4:21 5:0 => ()
expr @ 5:25 6:0 => ()
expr @ 7:22 8:0 => ()
expr @ 10:29 10:31 => i32
expr @ 10:19 10:33 => {value: i32}
expr @ 11:3 11:7 => fn free(...)
expr @ 11:9 11:10 => V<i32>
expr @ 11:8 11:10 => V<i32>*
expr @ 11:3 11:11 => ()
expr @ 12:13 12:16 => fn foo(...)
expr @ 12:18 12:19 => V<i32>
expr @ 12:17 12:19 => V<i32>*
expr @ 12:13 12:20 => i32*
expr @ 13:6 13:9 => fn foo(...)
expr @ 13:11 13:12 => V<i32>
expr @ 13:10 13:12 => V<i32>*
expr @ 13:6 13:13 => i32*
expr @ 14:3 14:4 => V<i32>
expr @ 14:9 14:11 => i32
expr @ 14:3 14:12 => ()
expr @ 15:13 15:14 => V<i32>
expr @ 15:12 15:14 => V<i32>*
expr @ 15:12 15:21 => i32*
expr @ 15:10 15:21 => i32
expr @ 15:3 15:21 => never
expr @ 9:16 17:1 => ()



define void @root$free$LT$i32$GT$({i32}* %lt.0) {
prologue:
%l.0 = alloca {i32}*
store {i32}* %lt.0, {i32}** %l.0
br label %bb.0

bb.0:
ret void

}

define i32* @root$V$foo$LT$i32$GT$({i32}* %lt.1) {
prologue:
%l.0 = alloca i32*
%l.1 = alloca {i32}*
store {i32}* %lt.1, {i32}** %l.1
br label %bb.0

bb.0:
%t.0 = load {i32}*, {i32}** %l.1
%t.1 = getelementptr {i32}, {i32}* %t.0, i32 0, i32 0
%l.2 = bitcast i32* %t.1 to i32*
store i32* %l.2, i32** %l.0
%ret.2 = load i32*, i32** %l.0
ret i32* %ret.2

}

define void @root$V$bar$LT$i32$$i32$GT$({i32} %lt.0, i32 %lt.1) {
prologue:
%l.0 = alloca {i32}
%l.1 = alloca i32
store {i32} %lt.0, {i32}* %l.0
store i32 %lt.1, i32* %l.1
br label %bb.0

bb.0:
ret void

}

define i32 @main() {
prologue:
%l.0 = alloca i32
%l.1 = alloca {i32}
br label %bb.0

bb.0:
%t.0 = alloca {i32}
%p.1 = getelementptr {i32}, {i32}* %t.0, i32 0, i32 0
store i32 42, i32* %p.1
%t.2 = load {i32}, {i32}* %t.0
store {i32} %t.2, {i32}* %l.1
%l.3 = bitcast {i32}* %l.1 to {i32}*
%l.2 = call {} bitcast(void({i32}*)* @root$free$LT$i32$GT$ to {}({i32}*)*)({i32}* %l.3)
br label %bb.1

bb.1:
%l.5 = bitcast {i32}* %l.1 to {i32}*
%l.4 = call i32* @root$V$foo$LT$i32$GT$({i32}* %l.5)
br label %bb.2

bb.2:
%l.7 = bitcast {i32}* %l.1 to {i32}*
%l.6 = call i32* @root$V$foo$LT$i32$GT$({i32}* %l.7)
br label %bb.3

bb.3:
%l.8 = load {i32}, {i32}* %l.1
%l.9 = call {} bitcast(void({i32}, i32)* @root$V$bar$LT$i32$$i32$GT$ to {}({i32}, i32)*)({i32} %l.8, i32 42)
br label %bb.4

bb.4:
%l.10 = bitcast {i32}* %l.1 to {i32}*
%l.11 = call i32* @root$V$foo$LT$i32$GT$({i32}* %l.10)
br label %bb.5

bb.5:
%t.3 = alloca i32*
store i32* %l.11, i32** %t.3
%t.4 = load i32*, i32** %t.3
%l.12 = load i32, i32* %t.4
store i32 %l.12, i32* %l.0
%ret.5 = load i32, i32* %l.0
ret i32 %ret.5

}


